scio-core/src/main/scala/com/spotify/scio/ScioContext.scala

```diff
@@ -180,8 +180,11 @@ object ScioContext {
 
   /** Create a new [[ScioContext]] instance for testing. */
   def forTest(): ScioContext = {
+    // when running unit tests force 1 thread per job,
+    // otherwise each JobTest spawns as many threads as CPU cores available
+    // this is intended to make tests faster
     val opts = PipelineOptionsFactory
-      .fromArgs("--appName=" + TestUtil.newTestId())
+      .fromArgs("--targetParallelism=1", "--appName=" + TestUtil.newTestId())
```

Comment:
Some concurrency bugs won't be revealed with single thread though. OTOH it will make those tests less flaky.
Should we make this configurable? As an env var, `sys.props` or optional argument to `forTest` & `JobTest`.

Reply:
> Should we make this configurable? As an env var, sys.props or optional argument to forTest & JobTest.

Good idea. I will think about that.

Reply:
So, I realized there are at least 3 places that we might need this logic. `ScioContext`, `JobTest` and [Flo's `ScioOperator`](https://github.com/spotify/flo/blob/master/contrib/flo-scio_2.12/src/main/scala/com/spotify/flo/contrib/scio/ScioOperator.scala#L64). I am trying to think if there is a way to simplify and/or avoid duplication.

Also, what do we want as default?

- If we choose `1`, this repo and all scio users will easily get faster and less flaky builds. But they might not notice concurrency bugs.
- If we make this optional, benefits are opt in, we need to instruct users on how to use it.

Reply:
Alternatively we can configure scalatest to run tests sequentially instead of parallel?
Not sure how much users really benefit from this given that most of our repos are smallish and we build on our own Jenkins/VMs which are more customizable. It'll probably benefit Scio more given the amount of tests and tiny CircleCI containers, but for that we can opt in.
@regadas thoughts?

Reply:
If it were an optional arg in the context that would be nice. You should be able to run unit tests in a single-threaded environment so that you can get in-process step-through capabilities on methods that take SCollections as args. Not sure why you'd need single-threadedness when you're running in any context other than a debugger though.

Reply:
> If it were an optional arg in the context that would be nice. You should be able to run unit tests in a single-threaded environment so that you can get in-process step-through capabilities on methods that take SCollections as args.

Yes. I can look into that.

> Alternatively we can configure scalatest to run tests sequentially instead of parallel?
> Not sure why you'd need single-threadedness when you're running in any context other than a debugger though.

Scalatest already runs tests in parallel. I think this is good, because on unit tests that seems the right place to have parallelism. Scio tests generally do not need to run in parallel, as the data being tests is generally small (maybe except property based testing?). So I am in favor of parallel scalatest with single threaded scio test. We could also tests and compare to check what is best (I probably won't have time to help on that).